File: algorithms/python/longestWell-PerformingInterval/longestWell-PerformingInterval.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/longest-well-performing-interval/
# Author: Miao Zhang
# Date:   2021-04-12

class Solution:
    def longestWPI(self, hours: List[int]) -> int:
        presum = 0
        res = 0
        seen = {}
        for i, h in enumerate(hours):
            presum += (1 if h > 8 else -1)
            if presum not in seen:
                seen[presum] = i
            if presum > 0:
                res = i + 1
            elif presum - 1 in seen:
                res = max(res, i - seen[presum - 1])
        return res

File: algorithms/cpp/longestWell-PerformingInterval/longestWell-PerformingInterval.cpp
// Source: https://leetcode.com/problems/longest-well-performing-interval/
// Author: Miao Zhang
// Date:   2021-04-12

class Solution {
public:
    int longestWPI(vector<int>& hours) {
        unordered_map<int, int> idx;
        int sums = 0;
        int res = 0;
        for (int i = 0; i < hours.size(); i++) {
            sums += hours[i] > 8 ? 1: -1;
            if (sums > 0) {
                res = i + 1;
            } else {
                if (!idx.count(sums)) {
                    idx[sums] = i;
                }
                if (idx.count(sums - 1)) {
                    res = max(res, i - idx[sums - 1]);
                }
            }
        }
        return res;
    }
};
